SAM III.5 Feature Branch - GAWB-2526

gpolumbo-broad opened the pull request:
Notes to Reviewers:

The heart of logic for this issue is in `ManagedGroupService.scala`.  All of the methods for this class could have easily been implemented inside `ResourceService.scala`, however, for fear of bloating that class, @UrsaStutsman and I decided to implement a separate class.  

As a result, this required me to change the parameter list of `SamRoutes` constructor, and therefore the `TestSamRoutes` constructor. 

The concern here is:  **_What happens then next time we add a new "special" resource-type?_**  Today it's ManagedGroups, tomorrow Workspaces, etc.  Do we want to be in the habit of bloating the parameter list of `SamRoutes` every time we need to add a resource-type?  

For now, the `ManagedGroupService` is composed with a `ResourceService` and a Map of Resource Types (this map is passed in only because that field is marked `private` in the `ResourceService` class).  This works fine...for now.  But I think we have other options:

1.  Encapsulate the `ManagedGroupService` (and its construction) inside `ResourceService`. 
1.  Create a `ManagedGroupService` factory method in `ResourceService`
1.  Make the `ManagedGroupService` into a trait and mix it into `ResourceService` (not sure how to make this work given that `ManagedGroupService` necessarily relies on methods in `ResourceService` but it looks like there's some wizardry with `type val` that might get us what we need?)
1.  Use Implicits to "inject" the methods from `ManagedGroupService` into `ResourceService` (Admittedly, I've only done a little bit of reading on this and I may be totally off-base with this suggestion)
1.  Keep it as implemented
1.  Other ideas?

My goals are:
1.  Make the construction of `SamRoutes` as easy as possible by not requiring us to expand the parameter list on the constructor every time we add a new special subject type.
1.  Ensure that we can keep logic for managing special resource types in their own classes
1.  Implement a solution that follows team best practices and coding standards

One final question: 
1. `ResourceService.resourceTypes` is `private`.  I'm not sure why.  If we make it public (which seems harmless) then we don't need to pass it into the `ManagedGroupService` constructor. 

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Make sure Swagger is updated if API changes
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

gpolumbo-broad committed:
GAWB-2526 pending tests and first stabs at getting the routes created

gpolumbo-broad committed:
GAWB-2526 Added ManagedGroupService class and did some refactoring to ResourceService to support reuse of methods by ManagedGroupService

gpolumbo-broad committed:
GAWB-2526 Tests for ManagedGroupService

gpolumbo-broad committed:
GAWB-2526 Tests for ManagedGroupService create and delete

gpolumbo-broad committed:
GAWB-2526 Tests for ManagedGroupRoutes and also got swagger urls added

gpolumbo-broad committed:
GAWB-2526 Tests for ManagedGroupRoutes.  Identified some todo items for requirements clarification.

TODO items for improving tests of AccessPolicyDAO.deleteResource

gpolumbo-broad committed:
GAWB-2526 Users without the Delete permission are not allowed to delete Managed Groups

gpolumbo-broad committed:
GAWB-2526 Taking care of TODOs

gpolumbo-broad committed:
GAWB-2526 Implemented simple validations for managed group email

gpolumbo-broad committed:
GAWB-2526 Cleaned up TODOs

gpolumbo-broad committed:
Merge branch 'develop' into gpolumbo-GAWB-2526

# Conflicts:
#	src/main/resources/reference.conf
#	src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala

gpolumbo-broad committed:
GAWB-2526 merged changes from develop branch

dvoet commented on src/main/resources/swagger/api-docs.yaml:
```diff
@@ -157,6 +157,66 @@ paths:
+      summary: Show details of the group
+      responses:
+        200:
+          description: Group found
```
need the schema for a 200 response

dvoet commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutes.scala:
```diff
@@ -0,0 +1,64 @@
+  }
+
+  private def handlePostGroup(groupId: String, userInfo: UserInfo) = {
+    complete(managedGroupService.createManagedGroup(ResourceId(groupId), userInfo).map(_ => StatusCodes.NoContent))
```
this should return status code Created

gpolumbo-broad replied:
fixed

dvoet commented on src/main/resources/swagger/api-docs.yaml:
```diff
@@ -157,6 +157,66 @@ paths:
+      responses:
+        204:
+          description: Group created
+        500:
```
should also return 409 if the group already exists

dvoet replied:
and 400 if the group name is invalid

gpolumbo-broad replied:
fixed

gpolumbo-broad committed:
GAWB-2526 basic description fix and removing incorrect comment on a test

dvoet commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutes.scala:
```diff
@@ -0,0 +1,64 @@
+
+  private def handleDeleteGroup(groupId: String, userInfo: UserInfo) = {
+    val resource = Resource(ManagedGroupService.ManagedGroupTypeName, ResourceId(groupId))
+    requireOneOfAction(resource, Set(SamResourceActions.delete), userInfo) {
```
use the short cut `requireAction` instead of `requireOneOfAction`

gpolumbo-broad replied:
fixed

dvoet commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupService.scala:
```diff
@@ -0,0 +1,89 @@
+  def managedGroupType: ResourceType = resourceTypes.getOrElse(ManagedGroupService.ManagedGroupTypeName, throw new WorkbenchException(s"resource type ${ManagedGroupService.ManagedGroupTypeName.value} not found"))
+  def memberRole = managedGroupType.roles.find(_.roleName == ManagedGroupService.MemberRoleName).getOrElse(throw new WorkbenchException(s"${ManagedGroupService.MemberRoleName} role does not exist in $managedGroupType"))
+
+  // Important Note, see: org/broadinstitute/dsde/workbench/sam/model/SamModel.scala:90
```
I guarantee this comment will be out of date

breilly2 replied:
Also what is the significance in the context of `createManagedGroup`?

gpolumbo-broad replied:
fixed

dvoet commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupService.scala:
```diff
@@ -0,0 +1,89 @@
+object ManagedGroupService {
+  val MemberRoleName = ResourceRoleName("member")
+  val ManagedGroupTypeName = ResourceTypeName("managed-group")
+  val LocalPartRe = "^[A-z0-9_-]+$".r
```
vals should start with lower case letter

gpolumbo-broad replied:
@dvoet my intention was that these should be treated like constants or `static final` values: https://docs.scala-lang.org/style/naming-conventions.html#constants-values-variable-and-methods

If you still want them changed to lowercase, happy to do so, just say so.

dvoet commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupService.scala:
```diff
@@ -0,0 +1,89 @@
+    ManagedGroupService.LocalPartRe.findFirstMatchIn(str) match {
+      case Some(_) => None
+      case None => Option(ErrorReport(s"You have specified a group name that contains characters that are not permitted in an email address. Group name may only contain alphanumeric characters, underscores, and dashes"))
+    }
```
regexes can be used in pattern matches so this also works

```
    str match {
      case ManagedGroupService.LocalPartRe() => None
      case _ => Option(ErrorReport(s"You have specified a group name that contains characters that are not permitted in an email address. Group name may only contain alphanumeric characters, underscores, and dashes")) 
    }
```

gpolumbo-broad replied:
fixed

dvoet commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupService.scala:
```diff
@@ -0,0 +1,89 @@
+  }
+
+  private def generateManagedGroupEmail(resourceId: ResourceId): WorkbenchEmail = {
+    val localPart = resourceId.value
```
I prefer the term groupName instead of localPart

gpolumbo-broad replied:
fixed

gpolumbo-broad commented:
jenkins retest

dvoet commented on src/test/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutesSpec.scala:
```diff
@@ -0,0 +1,97 @@
+    assertCreateGroup(samRoutes)
+  }
+
+  "GET /api/group/{groupName}" should "return a flattened list of users who are in this group" in {
```
test name is wrong

gpolumbo-broad replied:
fixed

dvoet commented on src/test/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutesSpec.scala:
```diff
@@ -0,0 +1,97 @@
+    Get(s"/api/group/$groupId") ~> samRoutes.route ~> check {
+      val expectedResource = Resource(ManagedGroupService.ManagedGroupTypeName, ResourceId(groupId))
+      status shouldEqual StatusCodes.OK
+      responseAs[String] should include (s"${groupId}@${samRoutes.resourceService.emailDomain}")
```
In general, route tests should test return codes and input/output formats but not response values. The service level tests should test that output is actually as expected.

gpolumbo-broad replied:
fixed

dvoet commented on src/test/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupServiceSpec.scala:
```diff
@@ -0,0 +1,123 @@
+    intercept[WorkbenchExceptionWithErrorReport] {
+      makeGroup(groupName)
+    }
+    runAndWait(managedGroupService.loadManagedGroup(resourceId)).isDefined shouldEqual false
```
drop .isDefined then make it  `shouldEqual None`

gpolumbo-broad replied:
fixed

dvoet commented on src/test/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupServiceSpec.scala:
```diff
@@ -0,0 +1,123 @@
+
+  "ManagedGroupService get" should "return the Managed Group resource" in {
+    makeGroup()
+    runAndWait(managedGroupService.loadManagedGroup(resourceId)).isDefined shouldEqual true
```
test that the actual value returned is what is expected:
`runAndWait(managedGroupService.loadManagedGroup(resourceId)) shouldEqual Some(....)`

gpolumbo-broad replied:
fixed

dvoet commented on src/test/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupServiceSpec.scala:
```diff
@@ -0,0 +1,123 @@
+    runAndWait(managedGroupService.loadManagedGroup(resourceId)).isDefined shouldEqual true
+  }
+
+  // NOTE: All since we don't have a way to look up policies directly without going through a Resource, this test
```
policies are sub contexts of resources so this is not possible give the current ldap structure

breilly2 commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutes.scala:
```diff
@@ -0,0 +1,64 @@
+package org.broadinstitute.dsde.workbench.sam.api
+
+// package object for implicit ErrorReportSource
```
This goes along with line 5? Guessing IntelliJ shifted things around during an auto-import.

breilly2 commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutes.scala:
```diff
@@ -0,0 +1,64 @@
+  }
+
+  private def handleGetGroup(groupId: String) = {
+    complete (
```
Should we protect this so that only members, or even only owners, can read a group?

gpolumbo-broad replied:
@breilly2 excellent question.  @dvoet and I talked about this a bit and my understanding was that the decision for now was to leave it as is since we aren't really revealing any information and somebody could "guess" if a group exists simply by trying to POST groups with names they're looking for.

gpolumbo-broad committed:
GAWB-2526 fixing Doug's PR comments

gpolumbo-broad committed:
GAWB-2526 fixing Doug's PR comments

gpolumbo-broad committed:
GAWB-2526 fixing Doug's PR comments

gpolumbo-broad committed:
GAWB-2526 fixing Doug's PR comments - assertions about content of the response are in the SericeSpec.  The RoutesSpec only asserts the value of the response codes

gpolumbo-broad committed:
GAWB-2526 fixing Doug's PR comments

dvoet commented on src/main/resources/swagger/api-docs.yaml:
```diff
@@ -157,6 +157,68 @@ paths:
+    get:
+      summary: Show email address of the group
+      responses:
+        200:
```
still need a schema for 200 response

gpolumbo-broad replied:
fixed

dvoet commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupService.scala:
```diff
@@ -0,0 +1,88 @@
+
+  // Per dvoet, when asking for a group, we will just return the group email
+  def loadManagedGroup(groupId: ResourceId): Future[Option[WorkbenchEmail]] = {
+    resourceService.directoryDAO.loadGroup(WorkbenchGroupName(groupId.value)).map(_.map(_.email))
```
I think I would prefer this service class taking directoryDAO as a constructor parameter rather than using resourceService's (which probably should be private)

gpolumbo-broad replied:
Checked with @dvoet and going to make all the values passed to the `ResourceService` constructor private, and will update the `ManagedGroupService` constructor to take all of the required components as parameters in the constructor.

gpolumbo-broad replied:
fixed

breilly2 commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupService.scala:
```diff
@@ -0,0 +1,88 @@
+  }
+}
+
+object ManagedGroupService {
```
Not sure where that other thread went, but I'll chime in and vote for initial lowercase as well. `val`s are immutable and are far more common than `var`s in scala so there isn't any need to call special attention to these. My brain is already rewired to read the initial uppercase as a type name rather than a val name.

gpolumbo-broad replied:
fixed

breilly2 commented on src/test/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutesSpec.scala:
```diff
@@ -0,0 +1,115 @@
+import org.scalatest.{BeforeAndAfter, FlatSpec, GivenWhenThen, Matchers}
+
+/**
+  * Created by dvoet on 6/7/17.
```
Who created this?

gpolumbo-broad replied:
fixed

dvoet commented:
@gpolumbo-broad regarding your design question - I like the ideas behind dependency injection: a component should be told its dependencies, it shouldn't know where to find them. We do that through our constructors. There are certainly other approaches we can use but this is the simplest. I will say bad words about implicits though.

If it is becoming too unwieldy we have created a "Application" object in the past that contains all of our service classes and just pass that around once.

breilly2 commented on src/test/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutesSpec.scala:
```diff
@@ -0,0 +1,115 @@
+  private val defaultRoles = Set(defaultOwnerRole, ResourceRole(ManagedGroupService.MemberRoleName, Set.empty))
+  private val managedGroupResourceType = ResourceType(ManagedGroupService.ManagedGroupTypeName, resourceActionPatterns, defaultRoles, ownerRoleName)
+  private val resourceTypes = Map(managedGroupResourceType.name -> managedGroupResourceType)
+  private val groupId = "foo"
```
I'm usually in favor of inlining these in the tests. Extracting common values implies that usages are somehow linked, that a change in one place necessitates a change in every other place. Since these are used across (independent) tests, this kind of link/interdependency is the opposite of what we want.

To put it another way, a change to the value here should require review of every test that uses it to make sure that the behavior being tested hasn't changed. While slightly more verbose, repeating the instantiation of the value forces that. (Sure, someone could search/replace but the PR would still explicitly declare that it is being changed for multiple test cases and enumerates exactly which test cases are affected.)

In contrast, `ownerRoleName` here is fine because it is a single unique fact about the component being tested rather than data for a test case. There might be some gray area in the values between `ownerRoleName` and `groupId`... I haven't fully analyzed them all... but we can chat in person if you have questions or counter arguments.

gpolumbo-broad replied:
@breilly2 I totally understand where you're coming from and I think it's definitely a tradeoff.  You'll see in some of the tests, I am creating some new instances of the objects I need, whereas in most of the tests, the defaults are just fine.  

That said, the spec files I wrote are now oddballs that do not conform to the coding conventions you've been using.  If the team would prefer that I rewrite these tests in that explicit style, I am happy to do so.

breilly2 reviewed (changes requested):
Just a few things. I might have a little more feedback but some of what's in this review are the more important points and I didn't want to further delay those suggestions. Let me know if you have questions.

gpolumbo-broad committed:
GAWB-2526 adding response code schema descriptions to swagger api doc

gpolumbo-broad committed:
GAWB-2526 refactored ManagedGroupService to take AccessPolicyDAO, DirectoryDAO, and emailDomain as parameters to constructor.

Made all ResourceService fields private per Doug

gpolumbo-broad committed:
GAWB-2526 removed comments from import statements

gpolumbo-broad committed:
GAWB-2526 lowercased name of val's in ManagedGroupService companion object

gpolumbo-broad committed:
GAWB-2526 fixed authorship comment on ManagedGroupRoutesSpec

gpolumbo-broad committed:
Merge branch 'develop' into gpolumbo-GAWB-2526

gpolumbo-broad commented:
jenkins retest

breilly2 commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupService.scala:
```diff
@@ -5,24 +5,26 @@ import com.typesafe.scalalogging.LazyLogging
   * Created by gpolumbo on 2/21/2018.
   */
-class ManagedGroupService(resourceService: ResourceService, val resourceTypes: Map[ResourceTypeName, ResourceType]) extends LazyLogging {
+//(private val resourceTypes: Map[ResourceTypeName, ResourceType], private val accessPolicyDAO: AccessPolicyDAO, private val directoryDAO: DirectoryDAO, private val cloudExtensions: CloudExtensions, private val emailDomain: String)
```
Nit: removed commented out code

gpolumbo-broad replied:
fixed

breilly2 commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ManagedGroupService.scala:
```diff
@@ -37,53 +39,53 @@ class ManagedGroupService(resourceService: ResourceService, val resourceTypes: M
+    val errors = validateGroupNamePattern(groupName) ++ validateGroupNameLength(constructEmail(groupName))
     if (errors.nonEmpty)
-      throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Cannot create valid email address from group name" , errors.toSeq))
+      throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "Cannot create valid email address with the group name provided" , errors.toSeq))
```
Minor improvement: include the group name in the error message.

gpolumbo-broad replied:
fixed.

breilly2 reviewed (approved):
Minor nits then 👍

gpolumbo-broad committed:
GAWB-2526 removed extraneous comment and updated exception message to include the string that caused the exception

gpolumbo-broad committed:
GAWB-3260 Need to stop for the weekend, so just pushing what I got working

gpolumbo-broad committed:
GAWB-3260 Managed Groups cannot be deleted if they are a member of at least 1 other WorkbenchGroup

gpolumbo-broad committed:
Merge branch 'gpolumbo-GAWB-3260' into gpolumbo-GAWB-2526

gpolumbo-broad commented:
For GAWB-3260, aside from the intended change, please also note the change to `ManagedGroupService.deleteManagedGroup()`.  I believe that this is the more appropriate implementation.

@dvoet - I have read the CONTRIBUTING.md, and I don't think there's anything additional that I need to do, but I'd appreciate your confirmation.

breilly2 commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/JndiDirectoryDAO.scala:
```diff
@@ -54,10 +54,23 @@ class JndiDirectoryDAO(protected val directoryConfig: DirectoryConfig)(implicit
 
+  // Forcibly delete a group regardless of its membership
+  // If you want to verify that the group is not a member of any other group prior to deletion, use safeDeleteGroup
   override def deleteGroup(groupName: WorkbenchGroupName): Future[Unit] = withContext { ctx =>
```
Instead of `safeDeleteGroup`/`deleteGroup`, I'd prefer `deleteGroup`/`forceDeleteGroup`, making "safe" the default behavior. That would make it harder to accidentally discover and call the wrong function when working with a `DirectoryDAO`. I haven't looked at remaining callers of the current `deleteGroup` to know if that would make that code read funny. Thoughts?

gpolumbo-broad replied:
Agreed that this would make more sense.  I contemplated just having the single `deleteGroup` method and rewriting it to always be "safe", but was concerned about breaking code that I might not understand.  Now that I look at it though, up until this PR, the only code calling `DirectoryDAO.deleteGroup()` has been tests.  

Therefore, since this is the first and only production use case for deleting a group so far, I would suggest we just have the one method for `deleteGroup` that does it in the "safe" way, and if we need a "force" version later, we can implement it then. 

I am going to make this change now, which may take a little bit of time to make sure all the tests stay green.

gpolumbo-broad replied:
Fixed

gpolumbo-broad committed:
GAWB-3260 reimplemented the previously existing DirectoryDAO.deleteGroup method.  It was never used in production code before, so its logic has been updated to always ensure that groups are not members of any other groups before deleting a group.

Most significant impact was to tests to make sure that we properly deconstruct and delete groups in the right order.

gpolumbo-broad committed:
GAWB-3260 reimplemented the previously existing DirectoryDAO.deleteGroup method.  It was never used in production code before, so its logic has been updated to always ensure that groups are not members of any other groups before deleting a group.

Most significant impact was to tests to make sure that we properly deconstruct and delete groups in the right order.

breilly2 reviewed (approved):
👍

gpolumbo-broad commented:
jenkins retest

dvoet commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/JndiDirectoryDAO.scala:
```diff
@@ -54,8 +54,13 @@ class JndiDirectoryDAO(protected val directoryConfig: DirectoryConfig)(implicit
-    ctx.unbind(groupDn(groupName))
+  override def deleteGroup(groupName: WorkbenchGroupName): Future[Unit] = {
+    listAncestorGroups(groupName).flatMap { ancestors =>
+      if (ancestors.nonEmpty)
```
style nit - please always use curly braces, even for 1 line if statements. It makes it easier to read and maintain

dvoet commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/JndiDirectoryDAO.scala:
```diff
@@ -54,8 +54,13 @@ class JndiDirectoryDAO(protected val directoryConfig: DirectoryConfig)(implicit
+  override def deleteGroup(groupName: WorkbenchGroupName): Future[Unit] = {
+    listAncestorGroups(groupName).flatMap { ancestors =>
+      if (ancestors.nonEmpty)
+        throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Conflict, s"group ${groupName.value} cannot be deleted because it is a member of at least 1 other group"))
```
I don't think it is functionally different but you probably want to return Future.failed instead of throwing the exception.

gpolumbo-broad replied:
Cool, good to know.  FWIW:  https://stackoverflow.com/questions/31194138/difference-between-returning-future-failedexception-and-throwing-an-exception

breilly2 replied:
The discussion on the answer is pretty interesting, in particular https://stackoverflow.com/questions/31194138/difference-between-returning-future-failedexception-and-throwing-an-exception#comment50481668_31195611

gpolumbo-broad committed:
GAWB-2526 Single line if and else statements should still use curly braces.

Failed Future should call Future.failed(new Exception...) instead of just throwing the exception

gpolumbo-broad committed:
GAWB-2529 Added route to sync managed-groups with Google Groups (#116)

gpolumbo-broad commented:
Merged GAWB-2529 (PR: https://github.com/broadinstitute/sam/pull/116) into this branch

MatthewBemis commented:
:+1: on the group sync stuff

dvoet commented on src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensionRoutes.scala:
```diff
@@ -98,30 +98,41 @@ trait GoogleExtensionRoutes extends ExtensionRoutes with UserInfoDirectives with
             }
           }
+        } ~
+        path("group" / Segment / "sync") { groupName =>
```
I thought we decided not to make this end point but to automatically sync the group

gpolumbo-broad committed:
Merge branch 'develop' into gpolumbo-GAWB-2526

gpolumbo-broad committed:
Merge branch 'develop' into gpolumbo-GAWB-2526

gpolumbo-broad committed:
GAWB-2526 Deleting a managed group should ensure that the google group is removed too

gpolumbo-broad committed:
Merge branch 'develop' into gpolumbo-GAWB-2526

gpolumbo-broad committed:
GAWB-2526 Automatically sync Managed Group with Google on Create (#118)

* GAWB-2526 When creating a Managed Group in SAM, it should automatically be sync'd with Google as part of the creation process

gpolumbo-broad committed:
Merge branch 'develop' into gpolumbo-GAWB-2526

gpolumbo-broad committed:
GAWB-2528 Endpoints for listing/overwriting Admin/Member policies for a Managed Group (#122)

* GAWB-2528 Functionality for listing and overwriting email addresses in the Admin and Member policies

* ManagedGroupService uses single methods for listing and overwriting policy members

* ManagedGroupService companion object has some static enum-like vals in them to try to ensure that we can't accidentally create or access Managed Group policies or roles other than "Admin" and "Member"

* JndiDirectoryDao now returns None when rather than throwing a NameNotFoundException when trying to load an email for a non-existent subject

* Added new route for adding an individual email to a policy in managed group

* Added new route for deleting an individual email to a policy in managed group

gpolumbo-broad committed:
Merge branch 'develop' into gpolumbo-GAWB-2526

gpolumbo-broad committed:
GAWB-3336 SAM /api/groups lists groups for user (#128)

* GAWB-3336 Added endpoint for /api/groups to list the Managed Groups that a user belongs to

gpolumbo-broad committed:
GAWB-2526 adding missing tests for GET /api/groups/{groupName} (#129)

* GAWB-2526 Adding missing tests for GET /api/group/{groupName}

gpolumbo-broad committed:
Merge branch 'develop' into gpolumbo-GAWB-2526
